tests/007-datagrams.cpp

```diff
@@ -279,15 +298,317 @@ namespace oxen::quic::test
+            std::vector<std::future<bool>> data_futures{n};
+
+            for (size_t i = 0; i < n; ++i)
+                data_futures[i] = data_promises[i].get_future();
+
+            std::promise<bool> tls_promise;
+            std::future<bool> tls_future = tls_promise.get_future();
+
+            gnutls_callback outbound_tls_cb =
+                    [&](gnutls_session_t, unsigned int, unsigned int, unsigned int, const gnutls_datum_t*) {
+                        log::debug(log_cat, "Calling client TLS callback... handshake completed...");
+
+                        tls_promise.set_value(true);
+                        return 0;
+                    };
+
+            dgram_data_callback recv_dgram_cb = [&](bstring) {
+                log::debug(log_cat, "Calling endpoint receive datagram callback... data received...");
+
+                try
+                {
+                    data_counter += 1;
+                    data_promises.at(index).set_value(true);
+                    index += 1;
+                }
+                catch (std::exception& e)
+                {
+                    throw std::runtime_error(e.what());
+                }
+            };
+
+            opt::enable_datagrams split_dgram{Splitting::ACTIVE, (int)bufsize};
+
+            opt::local_addr server_local{};
+            opt::local_addr client_local{};
+
+            auto server_tls = GNUTLSCreds::make("./serverkey.pem"s, "./servercert.pem"s, "./clientcert.pem"s);
+            auto client_tls = GNUTLSCreds::make("./clientkey.pem"s, "./clientcert.pem"s, "./servercert.pem"s);
+            client_tls->set_client_tls_policy(outbound_tls_cb);
+
+            auto server_endpoint = test_net.endpoint(server_local, split_dgram, recv_dgram_cb);
+            REQUIRE_NOTHROW(server_endpoint->listen(server_tls));
+
+            opt::remote_addr client_remote{"127.0.0.1"s, server_endpoint->local().port()};
+
+            auto client = test_net.endpoint(client_local, split_dgram);
+            auto conn_interface = client->connect(client_remote, client_tls);
+
+            REQUIRE(tls_future.get());
+
+            REQUIRE(server_endpoint->datagrams_enabled());
+            REQUIRE(client->datagrams_enabled());
+
+            REQUIRE(conn_interface->datagrams_enabled());
+            REQUIRE(conn_interface->packet_splitting_enabled());
+
+            auto max_size = conn_interface->get_max_datagram_size();
+
+            std::basic_string<uint8_t> good_msg{};
+            uint8_t v{0};
+
+            while (good_msg.size() < max_size)
+                good_msg += v++;
+
+            for (size_t i = 0; i < n; ++i)
+                conn_interface->send_datagram(std::basic_string_view<uint8_t>{good_msg});
+
+            for (auto& f : data_futures)
+                REQUIRE(f.get());
+
+            REQUIRE(data_counter == int(n));
+
+            auto server_ci = server_endpoint->get_all_conns(Direction::INBOUND).front();
+
+            REQUIRE(server_ci->last_cleared() == 0);
```

Comment:
This confirms that `last_cleared` was set to the expected value, but it doesn't confirm that the bucket was actually cleared.  In order to do that, we would need to have a way to make the connection send one half of a packet but not the other half, and then we'd need to send enough packets to loop back around do that packet number and make sure the other side of the connection gets the correct new packet rather than some frankensteined packet.

Reply:
keep scrolling